### PR TITLE
[bugfix] Prevent on_before_batch_transfer to be called twice

### LINKED ISSUE #9644: `on_before_batch_transfer` fails when batch is unpacked.
## 🐛 Bug

I am trying to do some unpacking on the `on_before_batch_transfer` and even though it works for some batches, it fails at some point, it seems it fails when changing from training to validation. This is with the code on `master`. The same happens for `transfer_batch_to_device` and `on_after_batch_transfer`.

### To Reproduce

https://colab.research.google.com/drive/1Eu3a8lj0KUPTNbwj95uvJ0mwwYAfXzF6?usp=sharing

### Expected behavior

The training loop should run properly unpacking the batch before transfer.

### Environment

* CUDA:
	- GPU:
		- Tesla K80
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.5.0dev
	- tqdm:              4.62.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.12
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9644

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/utilities/test_fetching.py [hunk 1]
```diff
@@ -392,3 +392,45 @@ def __init__(self) -> None:
     m = InvalidModel()
     with pytest.raises(MisconfigurationException, match="is incompatible with `truncated_bptt_steps > 0`."):
         trainer.fit(m)
+
+
+@RunIf(min_gpus=1)
+def test_on_before_batch_transfer_with_unpacking(tmpdir):
```
Comment (rohitgr7) [inline tests/utilities/test_fetching.py:398]: I think you should better check `_apply_batch_transfer_handler` or check the call count of all 3 hooks: `on_before_batch_transfer`, `transfer_batch_to_device`, `on_after_batch_transfer`.

File: tests/utilities/test_fetching.py [hunk 2]
```diff
@@ -392,3 +392,45 @@ def __init__(self) -> None:
...
+        def validation_step(self, batch, batch_idx):
+            x, y_true = batch
+            output = self.layer(x)
+            loss = self.loss(y_true, output)
+            return {"x": loss}
+
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, num_sanity_val_steps=0, gpus=1)
+    dm = BoringDataModule()
+    trainer.fit(TestModel(), datamodule=dm)
+    assert dm.has_called_on_before_batch_transfer
```
Comment (kaushikb11) [inline tests/utilities/test_fetching.py:436]: Wouldn't it still be `True` if it's called multiple times?

File: tests/utilities/test_fetching.py (hunk 1 above)
Comment (carmocca) [reply tests/utilities/test_fetching.py:398]: Yeah, let's unit-test this instead of running a `Trainer`

Comment (tchaton) [reply tests/utilities/test_fetching.py:398]: Done !

File: tests/utilities/test_fetching.py (hunk 2 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:436]: Done !

File: tests/utilities/test_fetching.py [hunk 3]
```diff
@@ -392,3 +392,58 @@ def __init__(self) -> None:
     m = InvalidModel()
     with pytest.raises(MisconfigurationException, match="is incompatible with `truncated_bptt_steps > 0`."):
         trainer.fit(m)
+
+
+@RunIf(min_gpus=1)
```
Comment (rohitgr7) [inline tests/utilities/test_fetching.py:397]: I don't think we need gpus for this one.

File: tests/utilities/test_fetching.py [hunk 4]
```diff
@@ -392,3 +392,58 @@ def __init__(self) -> None:
...
+def test_transfer_hooks_with_unpacking(tmpdir):
+
+    """This test asserts the `transfer_batch` hooks are called only once per batch and data are properly moved to
+    gpu."""
```
Comment (rohitgr7) [inline tests/utilities/test_fetching.py:401]:
```suggestion
    """This test asserts the `transfer_batch` hooks are called only once per batch."""
```
this is enough I guess. there are no assertions to test whether the data is moved to gpu or not.

File: tests/utilities/test_fetching.py [hunk 5]
```diff
@@ -392,3 +392,58 @@ def __init__(self) -> None:
...
+        def on_after_batch_transfer(self, batch, dataloader_idx: int):
+            self.count_called_on_after_batch_transfer += 1
+            return super().on_after_batch_transfer(batch, dataloader_idx)
+
+    class TestModel(BoringModel):
+        def training_step(self, batch, batch_idx):
+            x, y_true = batch
+            output = self.layer(x)
+            loss = self.loss(y_true, output)
+            return {"loss": loss}
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/utilities/test_fetching.py:436]: removed !

File: tests/utilities/test_fetching.py [hunk 6]
```diff
@@ -392,3 +392,58 @@ def __init__(self) -> None:
...
+
+        def validation_step(self, batch, batch_idx):
+            x, y_true = batch
+            output = self.layer(x)
+            loss = self.loss(y_true, output)
+            return {"x": loss}
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/utilities/test_fetching.py:442]: Done !

File: tests/utilities/test_fetching.py [hunk 7]
```diff
@@ -392,3 +392,58 @@ def __init__(self) -> None:
...
+
+    class RandomDictDataset(RandomDataset):
+        def __getitem__(self, index):
+            return {"x": self.data[index], "y_true": torch.ones((2,)), "other": torch.ones((1,))}
+
+    class BoringDataModule(LightningDataModule):
+
+        count_called_on_before_batch_transfer = 0
+        count_called_transfer_batch_to_device = 0
+        count_called_on_after_batch_transfer = 0
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/utilities/test_fetching.py:411]: There is no need here.

File: tests/utilities/test_fetching.py (hunk 4 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:401]: Done !

File: tests/utilities/test_fetching.py [hunk 8]
```diff
@@ -392,3 +392,52 @@ def __init__(self) -> None:
...
+    class TestModel(BoringModel):
+        def training_step(self, batch, batch_idx):
+            x, _ = batch
+            return super().training_step(x, batch_idx)
+
+        def validation_step(self, batch, batch_idx):
+            x, _ = batch
+            return super().validation_step(x, batch_idx)
+
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, num_sanity_val_steps=0, gpus=1)
```
Comment (rohitgr7) [inline tests/utilities/test_fetching.py:438]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, num_sanity_val_steps=0)
```

File: tests/utilities/test_fetching.py [hunk 9]
```diff
@@ -392,3 +392,52 @@ def __init__(self) -> None:
     m = InvalidModel()
     with pytest.raises(MisconfigurationException, match="is incompatible with `truncated_bptt_steps > 0`."):
         trainer.fit(m)
+
+
+def test_transfer_hooks_with_unpacking(tmpdir):
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:397]: The test seems to pass without the `fetching.py` change

Comment (tchaton) [reply tests/utilities/test_fetching.py:397]: When reverting the change, it fails. 

```py
tests/utilities/test_fetching.py F                                                                                                                                                                                                        [100%]

=================================================================================================================== FAILURES ====================================================================================================================
______________________________________________________________________________________________________ test_transfer_hooks_with_unpacking _______________________________________________________________________________________________________

tmpdir = local('/tmp/pytest-of-thomas/pytest-1901/test_transfer_hooks_with_unpac0')

    def test_transfer_hooks_with_unpacking(tmpdir):
    
        """This test asserts the `transfer_batch` hooks are called only once per batch."""
    
        class RandomDictDataset(RandomDataset):
            def __getitem__(self, index):
                return {"x": self.data[index], "y_true": torch.ones((2,)), "other": torch.ones((1,))}
    
        class BoringDataModule(LightningDataModule):
    
            count_called_on_before_batch_transfer = 0
            count_called_transfer_batch_to_device = 0
            count_called_on_after_batch_transfer = 0
    
            def train_dataloader(self):
                return DataLoader(RandomDictDataset(32, 2), batch_size=1)
    
            def val_dataloader(self):
                return DataLoader(RandomDictDataset(32, 2), batch_size=1)
    
            def on_before_batch_transfer(self, batch, dataloader_idx: int):
                self.count_called_on_before_batch_transfer += 1
                return batch["x"], batch["y_true"]
    
            def transfer_batch_to_device(self, *args, **kwargs):
                self.count_called_transfer_batch_to_device += 1
                return super().transfer_batch_to_device(*args, **kwargs)
    
            def on_after_batch_transfer(self, batch, dataloader_idx: int):
                self.count_called_on_after_batch_transfer += 1
                return super().on_after_batch_transfer(batch, dataloader_idx)
    
        class TestModel(BoringModel):
            def training_step(self, batch, batch_idx):
                x, _ = batch
                return super().training_step(x, batch_idx)
    
            def validation_step(self, batch, batch_idx):
                x, _ = batch
                return super().validation_step(x, batch_idx)
    
        trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, num_sanity_val_steps=0)
        dm = BoringDataModule()
>       trainer.fit(TestModel(), datamodule=dm)

tests/utilities/test_fetching.py:440: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pytorch_lightning/trainer/trainer.py:592: in fit
    self._call_and_handle_interrupt(self._fit_impl, model, train_dataloaders, val_dataloaders, datamodule)
pytorch_lightning/trainer/trainer.py:544: in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
pytorch_lightning/trainer/trainer.py:622: in _fit_impl
    self._run(model)
pytorch_lightning/trainer/trainer.py:1051: in _run
    self._dispatch()
pytorch_lightning/trainer/trainer.py:1130: in _dispatch
    self.accelerator.start_training(self)
pytorch_lightning/accelerators/accelerator.py:86: in start_training
    self.training_type_plugin.start_training(trainer)
pytorch_lightning/plugins/training_type/training_type_plugin.py:167: in start_training
    self._results = trainer.run_stage()
pytorch_lightning/trainer/trainer.py:1140: in run_stage
    return self._run_train()
pytorch_lightning/trainer/trainer.py:1176: in _run_train
    self.fit_loop.run()
pytorch_lightning/loops/base.py:114: in run
    self.advance(*args, **kwargs)
pytorch_lightning/loops/fit_loop.py:218: in advance
    self.epoch_loop.run(data_fetcher)
pytorch_lightning/loops/base.py:114: in run
    self.advance(*args, **kwargs)
pytorch_lightning/loops/epoch/training_epoch_loop.py:130: in advance
    batch = self.trainer.accelerator.batch_to_device(batch)
pytorch_lightning/accelerators/accelerator.py:168: in batch_to_device
    return model._apply_batch_transfer_handler(batch, device=device, dataloader_idx=dataloader_idx)
pytorch_lightning/core/lightning.py:265: in _apply_batch_transfer_handler
    batch = self.on_before_batch_transfer(batch, dataloader_idx)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.utilities.test_fetching.test_transfer_hooks_with_unpacking.<locals>.BoringDataModule object at 0x7f09a8f3d2e0>
batch = (tensor([[-0.6167,  0.1250,  0.9018,  1.6746,  0.1797, -1.1677,  0.5528,  1.2277,
          1.7385,  0.8121,  0.5910, ....4832, -0.7357,
         -0.3253,  1.0912,  1.4609,  0.1629,  0.9131,  0.8772, -0.6539,  0.2025]]), tensor([[1., 1.]]))
dataloader_idx = 0

    def on_before_batch_transfer(self, batch, dataloader_idx: int):
        self.count_called_on_before_batch_transfer += 1
>       return batch["x"], batch["y_true"]
E       TypeError: tuple indices must be integers or slices, not str

tests/utilities/test_fetching.py:419: TypeError
------------------------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------------------------
Epoch 0:  25%|██▌       | 1/4 [00:00<00:00, 250.60it/s, loss=0.466, v_num=0]
------------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------------
INFO:pytorch_lightning.utilities.distributed:GPU available: True, used: False
INFO:pytorch_lightning.utilities.distributed:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.distributed:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type   | Params
---------------------------------
0 | layer | Linear | 66    
---------------------------------
66        Trainable params
0         Non-trainable params
66        Total params
0.000     Total estimated model params size (MB)
--------------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------------
INFO     pytorch_lightning.utilities.distributed:trainer.py:1400 GPU available: True, used: False
INFO     pytorch_lightning.utilities.distributed:trainer.py:1403 TPU available: False, using: 0 TPU cores
INFO     pytorch_lightning.utilities.distributed:trainer.py:1406 IPU available: False, using: 0 IPUs
INFO     pytorch_lightning.callbacks.model_summary:model_summary.py:73 
  | Name  | Type   | Params
---------------------------------
0 | layer | Linear | 66    
---------------------------------
66        Trainable params
0         Non-trainable params
66        Total params
0.000     Total estimated model params size (MB)
=============================================================================================================== warnings summary ================================================================================================================
.venv/lib/python3.8/site-packages/_pytest/config/__init__.py:1183
  /home/thomas/pytorch-lightning/.venv/lib/python3.8/site-packages/_pytest/config/__init__.py:1183: PytestDeprecationWarning: The --strict option is deprecated, use --strict-markers instead.
    self.issue_config_time_warning(

tests/utilities/test_fetching.py::test_transfer_hooks_with_unpacking
  /home/thomas/pytorch-lightning/pytorch_lightning/trainer/trainer.py:1409: UserWarning: GPU available but not used. Set the gpus flag in your trainer `Trainer(gpus=1)` or script `--gpus=1`.
    rank_zero_warn(

tests/utilities/test_fetching.py::test_transfer_hooks_with_unpacking
  /home/thomas/pytorch-lightning/pytorch_lightning/trainer/data_loading.py:111: UserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 255 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
    rank_zero_warn(

tests/utilities/test_fetching.py::test_transfer_hooks_with_unpacking
  /home/thomas/pytorch-lightning/pytorch_lightning/trainer/data_loading.py:367: UserWarning: The number of training samples (2) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
    rank_zero_warn(

tests/utilities/test_fetching.py::test_transfer_hooks_with_unpacking
  /home/thomas/pytorch-lightning/pytorch_lightning/trainer/data_loading.py:111: UserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 255 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
    rank_zero_warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================================ short test summary info ============================================================================================================
FAILED tests/utilities/test_fetching.py::test_transfer_hooks_with_unpacking - TypeError: tuple indices must be integers or slices, not str
```py

Comment (carmocca) [reply tests/utilities/test_fetching.py:397]: You are right, I guess I didn't undo the change properly

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by ananyahjha93 — APPROVED** [review]

**Pull request merged**